### ISSUE TITLE
plugins.showroom: No playable streams found on online URL

### ISSUE THREAD

**Issue by taznica** [streamlink/streamlink#4397]
### Checklist

- [X] This is a plugin issue and not a different kind of issue
- [X] [I have read the contribution guidelines](https://github.com/streamlink/streamlink/blob/master/CONTRIBUTING.md#contributing-to-streamlink)
- [X] [I have checked the list of open and recently closed plugin issues](https://github.com/streamlink/streamlink/issues?q=is%3Aissue+label%3A%22plugin+issue%22)
- [X] [I have checked the commit log of the master branch](https://github.com/streamlink/streamlink/commits/master)

### Streamlink version

Latest stable release

### Description

The SHOWROOM plugin does not find streams in latest stable version(3.2.0), while it finds available streams in 3.1.1 on online(live streaming) URL.

#### Process
`$ streamlink https://www.showroom-live.com/(room_url_key)`

#### Expected  result
The plugin gets available streams list on online URL.
For example:
`Available streams: 144p (worst), 540p (best)`

#### Current result
The plugin can't find any playable streams on online URL.
`error: No playable streams found on this URL: https://www.showroom-live.com/(room_url_key)`

### Debug log

```text
https://gist.github.com/taznica/045e699da5ec888469767f7bf9c9f7bf
```

**Comment by bastimeyer** [comment]
> does not find streams in latest stable version(3.2.0), while it finds available streams in 3.1.1

This is the commit history of the plugin
https://github.com/streamlink/streamlink/commits/master/src/streamlink/plugins/showroom.py
and it was last changed in 5df21cdce885b23820cee32c2d32f461f421d78f via #4065, which was part of the 3.0.0 release.

Post the **exact URL without removing parts of it**.

The site is region locked for me, and I don't feel like looking up a proxy/VPN for this, but the channels I've checked were all working fine except for the region lock of course.

**Comment by taznica** [comment]
For example, the following URLs were online now(at the time of posting this reply), so I ran streamlink on them:
https://www.showroom-live.com/8f94b5277743
https://www.showroom-live.com/c4fb26129576
https://www.showroom-live.com/48_Manami_Ichikawa

in 3.1.1, I got available streams on URL:
`Available streams: 144p (worst), 360p (best)`

but in 3.2.0 I got no available streams on URL:
`No playable streams found on this URL: https://www.showroom-live.com/8f94b5277743`
`No playable streams found on this URL: https://www.showroom-live.com/c4fb26129576`
`No playable streams found on this URL: https://www.showroom-live.com/48_Manami_Ichikawa`

(We can get online lives list in https://www.showroom-live.com/onlive .)

**Comment by bastimeyer** [comment]
First stream is offline right now.

Second stream is restricted, but works in 3.2.0 and 3.1.1
```
$ streamlink 'https://www.showroom-live.com/c4fb26129576' best -l debug
[cli][debug] OS:         Linux-5.16.14-1-git-x86_64-with-glibc2.35
[cli][debug] Python:     3.10.2
[cli][debug] Streamlink: 3.2.0+5.ge52c247
[cli][debug] Requests(2.27.1), Socks(1.7.1), Websocket(1.2.3)
[cli][debug] Arguments:
[cli][debug]  url=https://www.showroom-live.com/c4fb26129576
[cli][debug]  stream=['best']
[cli][debug]  --loglevel=debug
[cli][debug]  --player=mpv
[cli][info] Found matching plugin showroom for URL https://www.showroom-live.com/c4fb26129576
[plugins.showroom][debug] {'is_live': 1, 'room_id': 395598, 'room': {'content_region_permission': 1, 'is_free': 0}}
[plugins.showroom][error] This stream is restricted
error: No playable streams found on this URL: https://www.showroom-live.com/c4fb26129576
```

```
$ streamlink 'https://www.showroom-live.com/c4fb26129576' best -l debug
[cli][debug] OS:         Linux-5.16.14-1-git-x86_64-with-glibc2.35
[cli][debug] Python:     3.10.2
[cli][debug] Streamlink: 3.1.1
[cli][debug] Requests(2.27.1), Socks(1.7.1), Websocket(1.2.3)
[cli][debug] Arguments:
[cli][debug]  url=https://www.showroom-live.com/c4fb26129576
[cli][debug]  stream=['best']
[cli][debug]  --loglevel=debug
[cli][debug]  --player=mpv
[cli][info] Found matching plugin showroom for URL https://www.showroom-live.com/c4fb26129576
[plugins.showroom][debug] {'is_live': 1, 'room_id': 395598, 'room': {'content_region_permission': 1, 'is_free': 0}}
[plugins.showroom][error] This stream is restricted
error: No playable streams found on this URL: https://www.showroom-live.com/c4fb26129576
```

As you can see in the plugin code, the content restriction is a simple HTTP header check just before the HLSStream is returned, which means that everything up to that point should be working correctly.
https://github.com/streamlink/streamlink/blame/e52c247d84fe641f42e233f3163e68288ec24b75/src/streamlink/plugins/showroom.py#L71-L74

The third link is offline as well, but embeds a youtube video, which the plugin doesn't support, so if that's the actual content, then the canonical youtube video URL needs to be used as streamlink's input URL.

**Comment by taznica** [comment]
> First stream is offline right now.

I'm sorry that first and third stream are offline. 
Online streams often go offline quickly.
We can get online lives list in https://www.showroom-live.com/onlive .


> Second stream is restricted, but works in 3.2.0 and 3.1.1

Regarding the execution results you showed, I don't think I can say that either version worked well...

Both showed
`error: No playable streams found on this URL: https://www.showroom-live.com/c4fb26129576`

but this is because `This stream is restricted` as you said.

I expected that the plugin gets available streams list such as:
`Available streams: 144p (worst), 540p (best)`

on **online URL** in **no restrictions area**, and this is achieved in my environment with 3.1.1.

However, this is not achieved in 3.2.0, the plugin can't find any playable streams on online URL.
`error: No playable streams found on this URL: https://www.showroom-live.com/xxxxxxxxxx`


I selected 10 online URL at this time from the online lives list (https://www.showroom-live.com/onlive) and ran 3.1.1 and 3.2.0 streamlink.
As a result, for ALL 10 URLs, available streams were found in 3.1.1, but not in 3.2.0.

Tested URLs:
- https://www.showroom-live.com/48_TANIGUCHI_MAHINA
- https://www.showroom-live.com/shijoluna227
- https://www.showroom-live.com/JKT48_Kathrina
- https://www.showroom-live.com/48_FUKUDA_AKARI
- https://www.showroom-live.com/a-light196
- https://www.showroom-live.com/48_Sorano_Uemi
- https://www.showroom-live.com/d29f35298918
- https://www.showroom-live.com/Arcanakitv_N005
- https://www.showroom-live.com/rere_sakura28
- https://www.showroom-live.com/reika0217

Example results:
**3.2.0**
```
$ streamlink --loglevel debug https://www.showroom-live.com/48_TANIGUCHI_MAHINA
[cli][debug] OS:         macOS 11.6
[cli][debug] Python:     3.10.2
[cli][debug] Streamlink: 3.2.0
[cli][debug] Requests(2.27.1), Socks(1.7.1), Websocket(1.3.1)
[cli][debug] Arguments:
[cli][debug]  url=https://www.showroom-live.com/48_TANIGUCHI_MAHINA
[cli][debug]  --loglevel=debug
[cli][info] Found matching plugin showroom for URL https://www.showroom-live.com/48_TANIGUCHI_MAHINA
error: No playable streams found on this URL: https://www.showroom-live.com/48_TANIGUCHI_MAHINA
```

**3.1.1**
```
$ streamlink --loglevel debug https://www.showroom-live.com/48_TANIGUCHI_MAHINA
[cli][debug] OS:         macOS 11.6
[cli][debug] Python:     3.10.2
[cli][debug] Streamlink: 3.1.1
[cli][debug] Requests(2.27.1), Socks(1.7.1), Websocket(1.3.1)
[cli][debug] Arguments:
[cli][debug]  url=https://www.showroom-live.com/48_TANIGUCHI_MAHINA
[cli][debug]  --loglevel=debug
[cli][info] Found matching plugin showroom for URL https://www.showroom-live.com/48_TANIGUCHI_MAHINA
[plugins.showroom][debug] {'is_live': 1, 'room_id': 99668, 'room': {'content_region_permission': 1, 'is_free': 0}}
[utils.l10n][debug] Language code: ja_JP
Available streams: 144p (worst), 360p (best)
```


> As you can see in the plugin code, ~~~~~

I'm sorry I only roughly read the code.
I thought it might be due to that the data was not retrieved up to here, and `return` here.
https://github.com/streamlink/streamlink/blame/e52c247d84fe641f42e233f3163e68288ec24b75/src/streamlink/plugins/showroom.py#L48-L49
Sorry if I'm looking in the wrong place, I will read the code.


> The third link is offline as well, but embeds a youtube video, ~~~~~

The embedded YouTube shown is one variation of what is shown when that live is offline as an advertisement. I do not want to download that YouTube, I want to download the live streaming when that live was online.

**Comment by bastimeyer** [comment]
Please post the result of `pip list` from both python environments you're running the different streamlink versions in. As said and as you can see in the [3.2.0 changelog](https://streamlink.github.io/changelog.html#streamlink-3-2-0-2022-03-05), there haven't been any changes in regards to the plugin code or HLS code, so this might be an issue with one of the dependencies, most notably `charset-normalizer`.

> I thought it might be due to that the data was not retrieved up to here, and `return` here.
> [`e52c247`/src/streamlink/plugins/showroom.py#L48-L49 (blame)](https://github.com/streamlink/streamlink/blame/e52c247d84fe641f42e233f3163e68288ec24b75/src/streamlink/plugins/showroom.py#L48-L49)

It's also possible that the site is returning different data for your region, which then breaks the validation schema and `None` is returned without the info or error log messages like in the other if-blocks. In that case I can't help.

**Comment by taznica** [comment]
Here are the results of `pip list`:

3.2.0
```

$ pip list
Package            Version
------------------ ---------
certifi            2021.10.8
charset-normalizer 2.0.12
idna               3.3
install            1.3.5
isodate            0.6.1
lxml               4.8.0
pip                22.0.4
pycountry          22.3.5
pycryptodome       3.14.1
PySocks            1.7.1
requests           2.27.1
setuptools         60.5.0
six                1.16.0
streamlink         3.2.0
urllib3            1.26.8
websocket-client   1.3.1
```

3.1.1
```
pip list
Package            Version
------------------ ---------
certifi            2021.10.8
charset-normalizer 2.0.12
idna               3.3
isodate            0.6.1
lxml               4.8.0
pip                22.0.4
pycountry          22.3.5
pycryptodome       3.14.1
PySocks            1.7.1
requests           2.27.1
setuptools         58.1.0
six                1.16.0
streamlink         3.1.1
urllib3            1.26.8
websocket-client   1.3.1
```

**Comment by back-to** [comment]
can't reproduce anything of this

**Comment by thisisgk** [comment]
I thought it would be better to make a comment here too.

Like #4414, this issue seems to only occur on macOS (of specific versions).

I chose one URL from https://www.showroom-live.com/onlive as guided by the OP and tried on macOS 10.12.6. 
I made an environment with Python 3.10.4 `venv`, did `pip install streamlink`, then overwrote older versions there with `pip install streamlink==3.1.1` etc.

Though not match to the OP about 3.1.1 (I have no idea why the difference), I get the following results:

On macOS 10.12.6,
streamlink 3.2.0, 3.1.1, 3.0.2: don't detect streams,
streamlink 3.0.1: detects streams,
modified streamlink 3.2.0 as in #4414: detects streams.

On macOS 10.15.7, streamlink 3.2.0 detects streams fine. (I didn't try other versions.)
